File: C++/1482. Minimum Number of Days to Make m Bouquets.cpp
class Solution {
public:
    int possible(vector<int>& bloomDay, int k, int mid) {
        int tot = 0;
        int count = 0;
        for (int i = 0; i < bloomDay.size(); i++) {
            if (bloomDay[i] <= mid) {
                count++;
                if (count == k) {
                    tot++;
                    count = 0;
                }
            } else {
                count = 0;
            }
        }
        return tot;
    }
    int minDays(vector<int>& bloomDay, int m, int k) {
           long long totalFlowers = (long long)m * (long long)k;
        if (bloomDay.size() < totalFlowers) {
            return -1;
        }
        int low = *min_element(bloomDay.begin(), bloomDay.end());
        int high = *max_element(bloomDay.begin(), bloomDay.end());
        int ans = high + 1;  
        while (low <= high) {
            int mid = (low + high) / 2;
            if (possible(bloomDay, k, mid) >= m) {
                ans = min(ans, mid);
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return ans;
    }
};

File: JAVA/1482. Minimum Number of Days to Make m Bouquets.java
class Solution {
    public static boolean isPossible(int[] bloomDay, int mid, int m, int k) {
        int adj = 0, bc = 0;
        for (int i = 0; i < bloomDay.length; i++) {
            if (bloomDay[i] <= mid) {
                adj++;
                if (adj == k) {
                    bc++;
                    adj = 0;
                    if (bc == m) {
                        return true;
                    }
                }
            } else {
                adj = 0;
            }
        }
        return false;
    }    
    public int minDays(int[] bloomDay, int m, int k) {
        int low = Integer.MAX_VALUE;
        int high = Integer.MIN_VALUE;
        
        for (int i = 0; i < bloomDay.length; i++) {
            if (bloomDay[i] < low) {
                low = bloomDay[i];
            }
            if (bloomDay[i] > high) {
                high = bloomDay[i];
            }
        }
        int res = -1;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (isPossible(bloomDay, mid, m, k)) {
                res = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return res;
    }
}
